File: backend/services/transaction_categorizer.py
"""
Transaction Categorizer Service
Uses deterministic rules with Sarvam fallback to categorize financial transactions
based on description and common Indian spending patterns.
"""

import os
import json
from typing import Optional, List, Dict
from dotenv import load_dotenv

load_dotenv()

# Category definitions with Indian-specific keywords
CATEGORY_KEYWORDS = {
    "Food & Dining": [
        "swiggy", "zomato", "restaurant", "cafe", "food", "dining", "lunch", "dinner",
        "breakfast", "biryani", "pizza", "burger", "chai", "coffee", "tea", "snacks",
        "mess", "canteen", "dhaba", "hotel", "eatsure", "dominos", "kfc", "mcdonalds",
        "subway", "haldiram", "barbeque", "freshmen", "behrouz", "starbucks", "burger king",
        "pizza hut", "bistro", "diner", "coffee day", "ccd", "tim hortons", "taco bell",
        "oven story", "faasos", "box8", "wow momo"
    ],
    "Groceries": [
        "bigbasket", "grofers", "blinkit", "zepto", "dmart", "reliance fresh",
        "more megastore", "grocery", "vegetables", "fruits", "kirana", "supermarket",
        "provision", "ration", "jiomart", "amazon fresh", "nature's basket", "spencer's",
        "daily needs", "milk", "dairy", "bakery", "meat", "fish"
    ],
    "Transportation": [
        "uber", "ola", "rapido", "metro", "bus", "railway", "irctc", "petrol", "diesel",
        "fuel", "parking", "toll", "fastag", "auto", "cab", "taxi", "flight", "air",
        "indigo", "spicejet", "vistara", "redbus", "makemytrip", "cleartrip", "goibibo",
        "yatra", "ixigo", "blue smart", "indriver", "rly", "rail", "shell", "hpcl", "bpcl", "ioc"
    ],
    "Shopping": [
        "amazon", "flipkart", "myntra", "ajio", "nykaa", "meesho", "snapdeal",
        "shopclues", "paytm mall", "tata cliq", "lifestyle", "westside", "pantaloons",
        "max", "h&m", "zara", "uniqlo", "decathlon", "croma", "reliance digital",
        "trends", "zudio", "nike", "adidas", "puma", "skechers", "levi's", "marks & spencer"
    ],
    "Entertainment": [
        "netflix", "prime video", "hotstar", "disney", "youtube", "spotify", "gaana",
        "jio saavn", "pvr", "inox", "bookmyshow", "cinema", "movie", "games", "pubg",
        "dream11", "fantasy", "subscription", "ott", "cinepolis", "apple music", "audible",
        "sony liv", "zee5", "gaming", "steam", "playstation", "xbox"
    ],
    "Utilities": [
        "electricity", "water", "gas", "broadband", "internet", "wifi", "jio", "airtel",
        "vi", "vodafone", "bsnl", "act fibernet", "tata sky", "dish tv", "dth",
        "mobile recharge", "postpaid", "prepaid", "phone bill", "bescom", "bwssb",
        "mahavitaran", "adhani", "torrent power", "billdesk", "razorpay"
    ],
    "Healthcare": [
        "hospital", "clinic", "doctor", "pharma", "pharmacy", "medicine", "apollo",
        "medplus", "netmeds", "1mg", "pharmeasy", "tata 1mg", "diagnostic", "lab",
        "pathology", "consultation", "health", "medical", "dental", "eye", "practo",
        "cult.fit", "gym", "fitness"
    ],
    "Education": [
        "school", "college", "university", "course", "udemy", "coursera", "unacademy",
        "byju", "vedantu", "books", "stationery", "tuition", "coaching", "exam",
        "fees", "library", "kindle", "skillshare"
    ],
    "Investment": [
        "mutual fund", "sip", "zerodha", "groww", "upstox", "angel", "stocks", "shares",
        "trading", "demat", "nse", "bse", "investment", "fd", "fixed deposit",
        "ppf", "nps", "bonds", "gold", "sovereign", "smallcase", "kuvera", "indmoney"
    ],
    "Insurance": [
        "insurance", "lic", "hdfc life", "icici pru", "max life", "term", "health insurance",
        "motor insurance", "policy", "premium", "policybazaar", "digit", "acko", "navi"
    ],
    "EMI & Loans": [
        "emi", "loan", "installment", "credit card", "bajaj", "hdfc", "icici", "sbi",
        "home loan", "car loan", "personal loan", "education loan", "bnpl", "simpl", "lazypay"
    ],
    "Salary & Income": [
        "salary", "wages", "income", "payroll", "credit", "inward", "received",
        "payment received", "freelance", "bonus", "incentive", "commission", "refund", "interest"
    ],
    "Transfer": [
        "transfer", "neft", "imps", "rtgs", "upi", "gpay", "phonepe", "paytm",
        "bhim", "self transfer", "fund transfer", "account transfer", "cred"
    ],
    "Rent & Housing": [
        "rent", "house rent", "pg", "hostel", "accommodation", "maintenance",
        "society", "apartment", "flat", "deposit", "caution", "nobroker"
    ],
    "Personal Care": [
        "salon", "spa", "parlour", "haircut", "beauty", "cosmetics", "skincare",
        "grooming", "urban company", "looks", "javed habib"
    ]
}


class TransactionCategorizer:
    """
    Categorizes transactions using:
    1. MerchantService rules (User-defined, highest priority)
    2. Rule-based keyword matching (Fallback)
    3. Sarvam AI (Final fallback)
    """
    
    def __init__(self):
        # Import services
        from .sarvam_service import sarvam_service
        from .merchant_service import merchant_service
        self.sarvam_service = sarvam_service
        self.merchant_service = merchant_service
    
    async def _merchant_rule_categorize(self, description: str) -> str | None:
        """
        Check user-defined merchant rules first (Highest Priority).
        """
        try:
            rule = await self.merchant_service.find_matching_rule(description)
            if rule:
                return rule.category
        except Exception as e:
            print(f"Merchant rule lookup failed: {e}")
        return None
    
    def _rule_based_categorize(self, description: str) -> str | None:
        """
        Try to categorize using keyword matching.
        """
        desc_lower = description.lower()
        
        for category, keywords in CATEGORY_KEYWORDS.items():
            for keyword in keywords:
                if keyword in desc_lower:
                    return category
        
        return None
    
    async def categorize_single(self, description: str, amount: float, tx_type: str) -> str:
        """
        Categorize a single transaction using the priority chain:
        1. User-defined Merchant Rules (Fastest, Highest Priority)
        2. Keyword Matching (Fast)
        3. AI Fallback (Slowest)
        """
        # 1. Check user-defined merchant rules FIRST (Fast Path)
        category = await self._merchant_rule_categorize(description)
        if category:
            return category
        
        # 2. Try rule-based keyword matching
        category = self._rule_based_categorize(description)
        if category:
            return category
        
        # Fall back to Sarvam if rules don't match
        if self.sarvam_service.is_configured:
            try:
                prompt = f"""Categorize this Indian financial transaction into ONE of these categories:
- Food & Dining
- Groceries
- Transportation
- Shopping
- Entertainment
- Utilities
- Healthcare
- Education
- Investment
- Insurance
- EMI & Loans
- Salary & Income
- Transfer
- Rent & Housing
- Personal Care
- Other

Transaction: "{description}"
Amount: ₹{amount}
Type: {tx_type}

Respond with ONLY the category name, nothing else."""
                
                response = await self.sarvam_service.simple_chat(
                    prompt,
                    "You are a transaction categorizer. Respond with only one category name from the provided list."
                )
                return response.strip()
            except Exception as e:
                print(f"AI categorization failed: {e}")
        
        # Default fallback
        return "Other"
    
    async def categorize_batch(self, transactions: List[Dict]) -> List[Dict]:
        """
        Categorize multiple transactions efficiently.
        Uses batch processing for AI calls.
        """
        results = []
        uncategorized = []
        
        # First pass: Rule-based categorization
        for i, tx in enumerate(transactions):
            desc = tx.get("description", "")
            category = self._rule_based_categorize(desc)
            
            if category:
                tx["category"] = category
                results.append(tx)
            else:
                uncategorized.append((i, tx))
        
        # Second pass: AI categorization for uncategorized
        if uncategorized and self.sarvam_service.is_configured:
            try:
                # Build batch prompt
                batch_items = "\n".join([
                    f"{i+1}. {tx.get('description', 'Unknown')} - ₹{tx.get('amount', 0)}"
                    for i, (_, tx) in enumerate(uncategorized)
                ])

                prompt = f"""Categorize these Indian financial transactions. For each, respond with ONLY the category name from this list:
Food & Dining, Groceries, Transportation, Shopping, Entertainment, Utilities, Healthcare, Education, Investment, Insurance, EMI & Loans, Salary & Income, Transfer, Rent & Housing, Personal Care, Other

Transactions:
{batch_items}

Respond with one category per line, in the same order:"""

                response = await self.sarvam_service.simple_chat(
                    prompt,
                    "You are a transaction categorizer. Respond with category names only, one per line.",
                )
                categories = response.strip().split("\n")
                
                for j, (orig_idx, tx) in enumerate(uncategorized):
                    if j < len(categories):
                        tx["category"] = categories[j].strip()
                    else:
                        tx["category"] = "Other"
                    results.append(tx)
                    
            except Exception as e:
                print(f"Batch AI categorization failed: {e}")
                for _, tx in uncategorized:
                    tx["category"] = "Other"
                    results.append(tx)
        else:
            for _, tx in uncategorized:
                tx["category"] = "Other"
                results.append(tx)
        
        return results
    
    def get_spending_by_category(self, transactions: List[Dict]) -> Dict[str, float]:
        """
        Aggregate spending by category.
        """
        spending = {}
        for tx in transactions:
            category = tx.get("category", "Other")
            amount = tx.get("amount", 0)
            tx_type = tx.get("type", "expense")
            
            if tx_type == "expense":
                spending[category] = spending.get(category, 0) + amount
        
        return dict(sorted(spending.items(), key=lambda x: x[1], reverse=True))


# Singleton instance
transaction_categorizer = TransactionCategorizer()

async def categorize_transaction(description: str, amount: float, tx_type: str = 'expense') -> str:
    """Wrapper function for easier imports"""
    return await transaction_categorizer.categorize_single(description, amount, tx_type)

File: frontend/wealthin_flutter/lib/core/services/transaction_categorizer.dart
import 'package:flutter/material.dart';
import '../theme/app_theme.dart';

/// Transaction Categorizer — RegExp-based pipeline with confidence scoring
/// Runs offline in Dart. Categories aligned with Indian spending patterns.
class TransactionCategorizer {

  static const String other = 'Other';

  // Each entry: category → list of regex patterns (compiled once)
  // Brand/merchant exact names have higher rank (checked first)
  static final Map<String, List<RegExp>> _categoryRegex = _buildRegex({
    'Food & Dining': [
      r'\b(swiggy|zomato|eatsure|faasos|box8|wow[ -]?momo|behrouz|oven[ -]?story)\b',
      r'\b(dominos?|kfc|mcdonalds?|subway|pizza[ -]?hut|burger[ -]?king|taco[ -]?bell)\b',
      r'\b(starbucks|ccd|coffee[ -]?day|bistro|diner|haldiram|barbeque[ -]?nation)\b',
      r'\b(restaurant|cafe|food|dining|lunch|dinner|breakfast|snack|canteen|dhaba|mess)\b',
      r'\b(biryani|pizza|burger|chai|coffee|tea)\b',
    ],
    'Groceries': [
      r'\b(bigbasket|grofers|blinkit|zepto|jiomart|amazon[ -]?fresh|nature.?s[ -]?basket)\b',
      r'\b(dmart|reliance[ -]?fresh|more[ -]?megastore|spencer.?s)\b',
      r'\b(grocery|groceries|vegetable|fruit|kirana|supermarket|provision|ration|dairy|bakery)\b',
      r'\b(milk|meat|fish|eggs?)\b',
    ],
    'Transportation': [
      r'\b(uber|ola|rapido|indriver|blue[ -]?smart)\b',
      r'\b(irctc|redbus|makemytrip|cleartrip|goibibo|yatra|ixigo)\b',
      r'\b(indigo|spicejet|vistara|air[ -]?india|go[ -]?air)\b',
      r'\b(fastag|petrol|diesel|fuel|hpcl|bpcl|ioc|shell)\b',
      r'\b(metro|bus|railway|rly|auto|cab|taxi|parking|toll)\b',
    ],
    'Shopping': [
      r'\b(amazon|flipkart|myntra|ajio|nykaa|meesho|snapdeal|shopclues|tata[ -]?cliq)\b',
      r'\b(lifestyle|westside|pantaloons|max|h&m|zara|uniqlo|levi.?s|marks?[ -]?spencer)\b',
      r'\b(decathlon|croma|reliance[ -]?digital|trends|zudio)\b',
      r'\b(nike|adidas|puma|skechers)\b',
      r'\b(shopping|purchase)\b',
    ],
    'Entertainment': [
      r'\b(netflix|prime[ -]?video|hotstar|disney|youtube|spotify|gaana|jio[ -]?saavn|apple[ -]?music|audible)\b',
      r'\b(pvr|inox|bookmyshow|cinepolis|cinema|movie)\b',
      r'\b(dream11|fantasy|pubg|steam|playstation|xbox|gaming)\b',
      r'\b(sony[ -]?liv|zee5|ott|subscription)\b',
    ],
    'Utilities': [
      r'\b(jio|airtel|vi|vodafone|bsnl|act[ -]?fibernet)\b',
      r'\b(tata[ -]?sky|dish[ -]?tv|dth)\b',
      r'\b(bescom|bwssb|mahavitaran|torrent[ -]?power|adani[ -]?electricity)\b',
      r'\b(electricity|water[ -]?bill|gas[ -]?bill|broadband|internet|wifi)\b',
      r'\b(mobile[ -]?recharge|postpaid|prepaid|phone[ -]?bill|billdesk)\b',
    ],
    'Healthcare': [
      r'\b(apollo|medplus|netmeds|1mg|pharmeasy|tata[ -]?1mg|practo|cult\.?fit)\b',
      r'\b(hospital|clinic|doctor|pharmacy|pharma|medicine|diagnostic|lab|patholog)\b',
      r'\b(dental|optical|gym|fitness|health|medical|consultation)\b',
    ],
    'Education': [
      r'\b(udemy|coursera|unacademy|byjus?|vedantu|skillshare|kindle)\b',
      r'\b(school|college|university|coaching|tuition|exam[ -]?fee)\b',
      r'\b(books?|stationer|library|course[ -]?fee)\b',
    ],
    'SIP / Mutual Fund': [
      r'\b(sip|systematic[ -]?investment|mutual[ -]?fund|zerodha|groww|upstox|angel[ -]?one)\b',
      r'\b(smallcase|kuvera|indmoney|paytm[ -]?money|coin[ -]?zerodha)\b',
    ],
    'Investment': [
      r'\b(ppf|nps|elss|sgb|sovereign[ -]?gold|scss|nsc|fixed[ -]?deposit|fd|rd)\b',
      r'\b(stocks?|shares?|trading|demat|nse|bse|bonds?|gold[ -]?fund)\b',
      r'\b(investment|invest)\b',
    ],
    'Insurance': [
      r'\b(lic|hdfc[ -]?life|icici[ -]?pru|max[ -]?life|term[ -]?plan|health[ -]?insurance)\b',
      r'\b(motor[ -]?insurance|policybazaar|digit[ -]?insurance|acko|navi)\b',
      r'\b(insurance|insured|premium[ -]?paid|policy)\b',
    ],
    'EMI & Loans': [
      r'\b(emi|equated[ -]?monthly|home[ -]?loan|car[ -]?loan|personal[ -]?loan|education[ -]?loan)\b',
      r'\b(bajaj[ -]?finance|hdfc[ -]?bank|icici[ -]?bank|sbi|kotak|axis[ -]?bank)\b',
      r'\b(bnpl|simpl|lazypay|installment|repayment|loan[ -]?repay)\b',
    ],
    'Salary & Income': [
      r'\b(salary|wages|payroll|stipend|remuneration)\b',
      r'\b(interest[ -]?credit|dividend|bonus|incentive|commission)\b',
      r'\b(refund|cashback|reward|payment[ -]?received)\b',
    ],
    'Transfer': [
      r'\b(neft|imps|rtgs|upi|gpay|google[ -]?pay|phonepe|paytm|bhim|cred)\b',
      r'\b(fund[ -]?transfer|self[ -]?transfer|account[ -]?transfer)\b',
    ],
    'Rent & Housing': [
      r'\b(house[ -]?rent|pg|paying[ -]?guest|hostel|accommodation|society[ -]?maintenance)\b',
      r'\b(nobroker|magicbricks|99acres|flat[ -]?rent|apartment[ -]?rent|deposit)\b',
      r'\b(rent)\b',
    ],
    'Personal Care': [
      r'\b(urban[ -]?company|looks|javed[ -]?habib|naturals[ -]?salon)\b',
      r'\b(salon|spa|parlour|haircut|beauty|cosmetics?|skincare|grooming)\b',
    ],
    'ATM / Cash': [
      r'\b(atm[ -]?withdrawal|cash[ -]?withdrawal|atm)\b',
    ],
    'Crypto': [
      r'\b(bitcoin|btc|ethereum|eth|crypto|wazirx|coindcx|binance|usdt)\b',
    ],
  });

  static Map<String, List<RegExp>> _buildRegex(Map<String, List<String>> raw) {
    return raw.map((cat, patterns) => MapEntry(
      cat,
      patterns.map((p) => RegExp(p, caseSensitive: false)).toList(),
    ));
  }

  // ─────────────────────────────────────────────────────────────────────────
  //  PUBLIC API
  // ─────────────────────────────────────────────────────────────────────────

  /// All supported category names
  static List<String> get categories => _categoryRegex.keys.toList();

  /// Categorize a transaction description, returning category + confidence.
  static CategorizeResult categorize(String description) {
    if (description.isEmpty) return CategorizeResult(category: other, confidence: 0.0);

    // Check each category in order; first-match wins within a category
    for (final entry in _categoryRegex.entries) {
      for (int i = 0; i < entry.value.length; i++) {
        if (entry.value[i].hasMatch(description)) {
          // Earlier patterns = brand-level match = higher confidence
          final confidence = i == 0 ? 1.0 : (i == 1 ? 0.9 : 0.75);
          return CategorizeResult(category: entry.key, confidence: confidence);
        }
      }
    }
    return CategorizeResult(category: other, confidence: 0.0);
  }

  /// Convenience: returns just the category string (backwards compatible)
  static String categorizeSimple(String description) =>
      categorize(description).category;

  /// Get icon for category
  static IconData getIcon(String category) {
    switch (category) {
      case 'Food & Dining':        return Icons.restaurant_rounded;
      case 'Groceries':            return Icons.local_grocery_store_rounded;
      case 'Transportation':       return Icons.directions_car_rounded;
      case 'Shopping':             return Icons.shopping_bag_rounded;
      case 'Entertainment':        return Icons.movie_rounded;
      case 'Utilities':            return Icons.bolt_rounded;
      case 'Healthcare':           return Icons.medical_services_rounded;
      case 'Education':            return Icons.school_rounded;
      case 'SIP / Mutual Fund':    return Icons.trending_up_rounded;
      case 'Investment':           return Icons.show_chart_rounded;
      case 'Insurance':            return Icons.security_rounded;
      case 'EMI & Loans':          return Icons.credit_card_rounded;
      case 'Salary & Income':      return Icons.account_balance_rounded;
      case 'Transfer':             return Icons.compare_arrows_rounded;
      case 'Rent & Housing':       return Icons.home_rounded;
      case 'Personal Care':        return Icons.face_rounded;
      case 'ATM / Cash':           return Icons.local_atm_rounded;
      case 'Crypto':               return Icons.currency_bitcoin_rounded;
      default:                     return Icons.category_rounded;
    }
  }

  /// Get theme-palette colour for category
  static Color getColor(String category) {
    switch (category) {
      case 'Food & Dining':        return AppTheme.saffron;
      case 'Groceries':            return AppTheme.success;
      case 'Transportation':       return AppTheme.peacockTeal;
      case 'Shopping':             return AppTheme.lotusPink;
      case 'Entertainment':        return const Color(0xFF9966CC);
      case 'Utilities':            return AppTheme.warning;
      case 'Healthcare':           return AppTheme.error;
      case 'Education':            return AppTheme.peacockLight;
      case 'SIP / Mutual Fund':    return AppTheme.successLight;
      case 'Investment':           return AppTheme.success;
      case 'Insurance':            return const Color(0xFF5C6BC0);
      case 'EMI & Loans':          return const Color(0xFFEF6C00);
      case 'Salary & Income':      return AppTheme.champagneGold;
      case 'Transfer':             return AppTheme.silverMist;
      case 'Rent & Housing':       return const Color(0xFF795548);
      case 'Personal Care':        return AppTheme.lotusPink;
      case 'ATM / Cash':           return const Color(0xFF78909C);
      case 'Crypto':               return const Color(0xFFF7931A);
      default:                     return AppTheme.silverMist;
    }
  }
}

/// Result of a categorization operation
class CategorizeResult {
  final String category;
  /// Confidence: 1.0 = brand-exact match, 0.75 = keyword match, 0.0 = unmatched
  final double confidence;

  const CategorizeResult({required this.category, required this.confidence});

  bool get isConfident => confidence >= 0.75;
  bool get isMatched   => category != TransactionCategorizer.other;
}
